### Question 1
#### Title: Country Literacy Rates Across Age and Gender
[Link to Data Source](https://www.kaggle.com/datasets/thedevastator/youth-adult-literacy-rates-in-2019?resource=download)
* Introduction: This data set collected information on literacy rates spanning across different countries and age groups to provide insights on which regions have high or low literacy rates, as well as how literacy rates have changed over time from the span of 2010 to 2018. 

### Question 2

#### Importing Packages and Reading csv file

In [2]:
import numpy as np
import pandas as pd

In [3]:
df = pd.read_csv('Literacy_rates.csv')
df

index                     Region      Country  Year    Age  Gender  \
0         0  Central and Southern Asia  Afghanistan  2011    15+  female   
1         1  Central and Southern Asia  Afghanistan  2011    15+    male   
2         2  Central and Southern Asia  Afghanistan  2011    15+   total   
3         3  Central and Southern Asia  Afghanistan  2011  15-24  female   
4         4  Central and Southern Asia  Afghanistan  2011  15-24    male   
...     ...                        ...          ...   ...    ...     ...   
4950   4950         Sub-Saharan Africa     Zimbabwe  2014  25-64    male   
4951   4951         Sub-Saharan Africa     Zimbabwe  2014  25-64   total   
4952   4952         Sub-Saharan Africa     Zimbabwe  2014    65+  female   
4953   4953         Sub-Saharan Africa     Zimbabwe  2014    65+    male   
4954   4954         Sub-Saharan Africa     Zimbabwe  2014    65+   total   

      Literacy rate  
0          0.176121  
1          0.454171  
2          0.317411  
3          0.321132  
4          0.618791  
...             ...  
4950       0.897255  
4951       0.887874  
4952       0.734258  
4953       0.927231  
4954       0.811269  

[4955 rows x 7 columns]

#### First Five Rows

In [4]:
df.head()

index                     Region      Country  Year    Age  Gender  \
0      0  Central and Southern Asia  Afghanistan  2011    15+  female   
1      1  Central and Southern Asia  Afghanistan  2011    15+    male   
2      2  Central and Southern Asia  Afghanistan  2011    15+   total   
3      3  Central and Southern Asia  Afghanistan  2011  15-24  female   
4      4  Central and Southern Asia  Afghanistan  2011  15-24    male   

   Literacy rate  
0       0.176121  
1       0.454171  
2       0.317411  
3       0.321132  
4       0.618791

#### Last Five Rows

In [5]:
df.tail()

index              Region   Country  Year    Age  Gender  Literacy rate
4950   4950  Sub-Saharan Africa  Zimbabwe  2014  25-64    male       0.897255
4951   4951  Sub-Saharan Africa  Zimbabwe  2014  25-64   total       0.887874
4952   4952  Sub-Saharan Africa  Zimbabwe  2014    65+  female       0.734258
4953   4953  Sub-Saharan Africa  Zimbabwe  2014    65+    male       0.927231
4954   4954  Sub-Saharan Africa  Zimbabwe  2014    65+   total       0.811269

#### Finding How Many Rows and Columns Are in the Dataset

In [6]:
df.shape

(4955, 7)

* This dataset has 4955 Rows and 7 Columns

#### Finding Column Names

In [7]:
df.columns

Index(['index', 'Region', 'Country', 'Year', 'Age', 'Gender', 'Literacy rate'], dtype='object')

For this dataset, the column names are:
   * index
   * Region
   * Country
   * Year
   * Age
   * Gender
   * Literacy Rate

### Question 3
* For this dataset:
    * The index, Year, and Literacy Rate columns are numerical 
    * The Region, Country, Age, and Gender columns are categorical
* To avoid confusion, I believe it would be best to convert the index column into an object datatype so that Python will not include such inputs numerically or rather delete it entirely as it does not provide necessary information. For the Year column, I also believe it is best to convert it to an object datatype so that Python would not try to use it numerically, since it is only indicating the year. 

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4955 entries, 0 to 4954
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          4955 non-null   int64  
 1   Region         4955 non-null   object 
 2   Country        4955 non-null   object 
 3   Year           4955 non-null   int64  
 4   Age            4955 non-null   object 
 5   Gender         4955 non-null   object 
 6   Literacy rate  4955 non-null   float64
dtypes: float64(1), int64(2), object(4)
memory usage: 271.1+ KB


### Question 4
#### Finding Statistical Information

In [9]:
df.describe()

index         Year  Literacy rate
count  4955.000000  4955.000000    4955.000000
mean   2477.000000  2013.987891       0.816947
std    1430.529622     2.746204       0.232560
min       0.000000  2010.000000       0.004025
25%    1238.500000  2011.000000       0.722507
50%    2477.000000  2014.000000       0.930676
75%    3715.500000  2016.000000       0.985078
max    4954.000000  2018.000000       1.000000

Based on these statistics, it tells me that there is at least one instance of 100% literacy rate and one instance of a .4% literacy rate foumd within the data. The 25%, 50% and 75% percentiles tell me that the data is slightly concentracted but still somewhat spread out, since the first quartile has a 21% difference from the median compared to the third quartile's 5% difference from the median. The standard deivation confirms this finding with a 23% standard deiviation rate, indicating a high level of dispersion within the dataset. 

### Question 5
#### Finding Missing Values

In [10]:
df.isnull().sum()

index            0
Region           0
Country          0
Year             0
Age              0
Gender           0
Literacy rate    0
dtype: int64

By using .isnull().sum(), Python tells me that there are no null values within the dataset

#### Finding Duplicated Data

In [11]:
print('Number of Duplicates:', df.duplicated().sum())

Number of Duplicates: 0


After using .duplicated().sum(), Python tell me that there are no duplicates in the set, making it easier to interpret the data. 

### Question 6
#### Adding Based on Existing Column

Due to the possibility that the literacy rate in decimal form would be hard to interpret after a while, I decided to add a Percentage column which would convert the long decimal into an easy readable format and another column called Literacy Health to use the Percentage Column as a basis to label the overall literacy percentage for each index.
* **BUT!!!** Before making any changes, I had to copy the DataFrame into different one titled <u>litr<u>

In [12]:
litr = df.copy()
litr

index                     Region      Country  Year    Age  Gender  \
0         0  Central and Southern Asia  Afghanistan  2011    15+  female   
1         1  Central and Southern Asia  Afghanistan  2011    15+    male   
2         2  Central and Southern Asia  Afghanistan  2011    15+   total   
3         3  Central and Southern Asia  Afghanistan  2011  15-24  female   
4         4  Central and Southern Asia  Afghanistan  2011  15-24    male   
...     ...                        ...          ...   ...    ...     ...   
4950   4950         Sub-Saharan Africa     Zimbabwe  2014  25-64    male   
4951   4951         Sub-Saharan Africa     Zimbabwe  2014  25-64   total   
4952   4952         Sub-Saharan Africa     Zimbabwe  2014    65+  female   
4953   4953         Sub-Saharan Africa     Zimbabwe  2014    65+    male   
4954   4954         Sub-Saharan Africa     Zimbabwe  2014    65+   total   

      Literacy rate  
0          0.176121  
1          0.454171  
2          0.317411  
3          0.321132  
4          0.618791  
...             ...  
4950       0.897255  
4951       0.887874  
4952       0.734258  
4953       0.927231  
4954       0.811269  

[4955 rows x 7 columns]

In [13]:
litr['Percentage'] = ((litr['Literacy rate'])*100).round()
litr

index                     Region      Country  Year    Age  Gender  \
0         0  Central and Southern Asia  Afghanistan  2011    15+  female   
1         1  Central and Southern Asia  Afghanistan  2011    15+    male   
2         2  Central and Southern Asia  Afghanistan  2011    15+   total   
3         3  Central and Southern Asia  Afghanistan  2011  15-24  female   
4         4  Central and Southern Asia  Afghanistan  2011  15-24    male   
...     ...                        ...          ...   ...    ...     ...   
4950   4950         Sub-Saharan Africa     Zimbabwe  2014  25-64    male   
4951   4951         Sub-Saharan Africa     Zimbabwe  2014  25-64   total   
4952   4952         Sub-Saharan Africa     Zimbabwe  2014    65+  female   
4953   4953         Sub-Saharan Africa     Zimbabwe  2014    65+    male   
4954   4954         Sub-Saharan Africa     Zimbabwe  2014    65+   total   

      Literacy rate  Percentage  
0          0.176121        18.0  
1          0.454171        45.0  
2          0.317411        32.0  
3          0.321132        32.0  
4          0.618791        62.0  
...             ...         ...  
4950       0.897255        90.0  
4951       0.887874        89.0  
4952       0.734258        73.0  
4953       0.927231        93.0  
4954       0.811269        81.0  

[4955 rows x 8 columns]

In [14]:
def lit (num):
    if num <= 20:
        return 'Illiterate'
    if num <= 40: 
        return 'Slightly Literate'
    if num <= 60:
        return 'Average Literacy'
    if num <= 80:
        return 'Good Literacy'
    if num <= 100:
        return 'Excellent Literacy'

In [15]:
litr['Literacy Health'] = litr['Percentage'].apply(lit)
litr

index                     Region      Country  Year    Age  Gender  \
0         0  Central and Southern Asia  Afghanistan  2011    15+  female   
1         1  Central and Southern Asia  Afghanistan  2011    15+    male   
2         2  Central and Southern Asia  Afghanistan  2011    15+   total   
3         3  Central and Southern Asia  Afghanistan  2011  15-24  female   
4         4  Central and Southern Asia  Afghanistan  2011  15-24    male   
...     ...                        ...          ...   ...    ...     ...   
4950   4950         Sub-Saharan Africa     Zimbabwe  2014  25-64    male   
4951   4951         Sub-Saharan Africa     Zimbabwe  2014  25-64   total   
4952   4952         Sub-Saharan Africa     Zimbabwe  2014    65+  female   
4953   4953         Sub-Saharan Africa     Zimbabwe  2014    65+    male   
4954   4954         Sub-Saharan Africa     Zimbabwe  2014    65+   total   

      Literacy rate  Percentage     Literacy Health  
0          0.176121        18.0          Illiterate  
1          0.454171        45.0    Average Literacy  
2          0.317411        32.0   Slightly Literate  
3          0.321132        32.0   Slightly Literate  
4          0.618791        62.0       Good Literacy  
...             ...         ...                 ...  
4950       0.897255        90.0  Excellent Literacy  
4951       0.887874        89.0  Excellent Literacy  
4952       0.734258        73.0       Good Literacy  
4953       0.927231        93.0  Excellent Literacy  
4954       0.811269        81.0  Excellent Literacy  

[4955 rows x 9 columns]

### Question 7
#### Filtering the Data
* For my analysis, I decided to filter the data using the Country, Age, Gender, and Percentage columns to see if I could find a pattern correlating to the country, age, and gender of a person influcing their literacy rate. As shown in the first two rows, there is a big disparity between the male and female literacy rates with females at 18 percent and males in the same age group at 45 percent. Through this pattern, I question how prevelant this disparity is across different countries within the dataset.

In [16]:
litr_filtered = litr[['Country', 'Age', 'Gender', 'Percentage']]
litr_filtered

Country    Age  Gender  Percentage
0     Afghanistan    15+  female        18.0
1     Afghanistan    15+    male        45.0
2     Afghanistan    15+   total        32.0
3     Afghanistan  15-24  female        32.0
4     Afghanistan  15-24    male        62.0
...           ...    ...     ...         ...
4950     Zimbabwe  25-64    male        90.0
4951     Zimbabwe  25-64   total        89.0
4952     Zimbabwe    65+  female        73.0
4953     Zimbabwe    65+    male        93.0
4954     Zimbabwe    65+   total        81.0

[4955 rows x 4 columns]

### Question 8 
#### Unique Variables
Based on the categorical columns Region, Country, Age, and Gender, the unique values that I see as important for grouping and analysis include the Country, Age, and Gender columns as it could lead to outside research into the cultural practices in the region that may influence one's literacy.

In [17]:
litr.nunique()

index              4955
Region                7
Country             155
Year                  9
Age                   4
Gender                3
Literacy rate      4835
Percentage          101
Literacy Health       5
dtype: int64

In [18]:
litr['Region'].unique()

array(['Central and Southern Asia', 'Eastern and South-Eastern Asia',
       'Europe and Northern America', 'Latin America and the Caribbean',
       'Northern Africa and Western Asia', 'Oceania',
       'Sub-Saharan Africa'], dtype=object)

In [19]:
litr['Country'].unique()

array(['Afghanistan', 'Bangladesh', 'Bhutan', 'India',
       'Iran (Islamic Republic of)', 'Kazakhstan', 'Kyrgyzstan',
       'Maldives', 'Nepal', 'Pakistan', 'Sri Lanka', 'Tajikistan',
       'Turkmenistan', 'Uzbekistan', 'Brunei Darussalam', 'Cambodia',
       'China', 'China, Macao Special Administrative Region', 'Indonesia',
       "Lao People's Democratic Republic", 'Malaysia', 'Mongolia',
       'Myanmar', 'Philippines', 'Singapore', 'Thailand', 'Timor-Leste',
       'Viet Nam', 'Albania', 'Belarus', 'Bosnia and Herzegovina',
       'Bulgaria', 'Croatia', 'Estonia', 'Greece', 'Hungary', 'Italy',
       'Latvia', 'Lithuania', 'Malta', 'Montenegro', 'Portugal',
       'Republic of Moldova', 'Romania', 'Russian Federation',
       'San Marino', 'Serbia', 'Slovenia', 'Spain',
       'The former Yugoslav Republic of Macedonia', 'Ukraine',
       'Antigua and Barbuda', 'Argentina', 'Aruba', 'Barbados',
       'Bolivia (Plurinational State of)', 'Brazil', 'Chile', 'Colombia',
       'C

In [20]:
litr['Age'].unique()

array(['15+', '15-24', '25-64', '65+'], dtype=object)

In [21]:
litr['Gender'].unique()

array(['female', 'male', 'total'], dtype=object)

### Question 9
#### Overall Summary

 Overall, one thing I found intresting about this dataset is the extensive amount of information provided, showing how thorough the creator was at compliing the information into a fairly easily read format. One pattern that I picked up on is the literacy disparitities between males and females across different countries. When returning to the dataset, I hope to delve into that pattern to see how prevalent it is across different countries and find out what this signifies about the accessibility of literature and developing literacy. I would also like to see if age plays a role into one's literacy, and possibly do outside research to gain surrounding context  behind the given literacy rates.  

In [22]:
group = df.groupby('Age')
print(type(group))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [23]:
group.get_group('15+')

index                     Region      Country  Year  Age  Gender  \
0         0  Central and Southern Asia  Afghanistan  2011  15+  female   
1         1  Central and Southern Asia  Afghanistan  2011  15+    male   
2         2  Central and Southern Asia  Afghanistan  2011  15+   total   
11       11  Central and Southern Asia  Afghanistan  2018  15+  female   
12       12  Central and Southern Asia  Afghanistan  2018  15+    male   
...     ...                        ...          ...   ...  ...     ...   
4932   4932         Sub-Saharan Africa     Zimbabwe  2011  15+    male   
4933   4933         Sub-Saharan Africa     Zimbabwe  2011  15+   total   
4943   4943         Sub-Saharan Africa     Zimbabwe  2014  15+  female   
4944   4944         Sub-Saharan Africa     Zimbabwe  2014  15+    male   
4945   4945         Sub-Saharan Africa     Zimbabwe  2014  15+   total   

      Literacy rate  
0          0.176121  
1          0.454171  
2          0.317411  
11         0.298052  
12         0.554755  
...             ...  
4932       0.877648  
4933       0.835827  
4943       0.882838  
4944       0.891854  
4945       0.886934  

[1260 rows x 7 columns]

In [29]:
litr.to_csv('Lit2.csv', index = False)